contracts/modules/elrond-wasm-module-governance/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+		}
+
+		let actions = self.proposals().get(proposal_id).actions;
+		let mut actions_as_multiarg = Vec::with_capacity(actions.len());
+
+		for action in actions {
+			actions_as_multiarg.push(action.into_multiarg());
+		}
+
+		actions_as_multiarg.into()
+	}
+
+	// private
+
+	fn require_payment_token_governance_token(&self) -> SCResult<()> {
+		require!(
+			self.call_value().token() == self.governance_token_id().get(),
+			"Only Governance token accepted as payment"
+		);
+		Ok(())
+	}
+
+	fn require_valid_proposal_id(&self, proposal_id: usize) -> SCResult<()> {
+		require!(
+			self.is_valid_proposal_id(proposal_id),
+			"Invalid proposal ID"
+		);
+		Ok(())
+	}
+
+	fn is_valid_proposal_id(&self, proposal_id: usize) -> bool {
+		proposal_id >= 1 && proposal_id <= self.proposals().len()
+	}
+
+	fn proposal_exists(&self, proposal_id: usize) -> bool {
+		self.is_valid_proposal_id(proposal_id) && !self.proposals().item_is_empty(proposal_id)
+	}
+
+	fn total_gas_needed(&self, actions: &[GovernanceAction<Self::BigUint>]) -> u64 {
+		let mut total = 0;
+		for action in actions {
+			total += action.gas_limit;
+		}
+
+		total
+	}
+
+	fn get_sc_balance(&self, token_id: &TokenIdentifier) -> Self::BigUint {
+		if token_id.is_egld() {
+			self.blockchain().get_sc_balance()
+		} else {
+			self.blockchain()
+				.get_esdt_balance(&self.blockchain().get_sc_address(), token_id, 0)
+		}
+	}
+
+	/// specific votes/downvotes are not cleared,
+	/// as they're used for reclaim tokens logic and cleared one by one
+	fn clear_proposal(&self, proposal_id: usize) {
+		self.proposals().clear_entry(proposal_id);
+		self.proposal_start_block(proposal_id).clear();
+		self.proposal_queue_block(proposal_id).clear();
+
+		self.total_votes(proposal_id).clear();
+		self.total_downvotes(proposal_id).clear();
+	}
+
+	// storage - general
+
+	#[storage_mapper("governance:proposals")]
+	fn proposals(&self) -> VecMapper<Self::Storage, GovernanceProposal<Self::BigUint>>;
+
+	/// Not stored under "proposals", as that would require deserializing the whole struct
+	#[storage_mapper("governance:proposalStartBlock")]
+	fn proposal_start_block(&self, proposal_id: usize) -> SingleValueMapper<Self::Storage, u64>;
+
+	#[storage_mapper("governance:proposalQueueBlock")]
+	fn proposal_queue_block(&self, proposal_id: usize) -> SingleValueMapper<Self::Storage, u64>;
+
+	#[storage_mapper("governance:votes")]
+	fn votes(&self, proposal_id: usize) -> MapMapper<Self::Storage, Address, Self::BigUint>;
+
+	#[storage_mapper("governance:downvotes")]
+	fn downvotes(&self, proposal_id: usize) -> MapMapper<Self::Storage, Address, Self::BigUint>;
+
+	/// Could be calculated by iterating over the "votes" mapper, but that costs a lot of gas
+	#[view(getTotalVotes)]
+	#[storage_mapper("governance:totalVotes")]
+	fn total_votes(&self, proposal_id: usize) -> SingleValueMapper<Self::Storage, Self::BigUint>;
+
+	/// Could be calculated by iterating over the "downvotes" mapper, but that costs a lot of gas
```

Comment:
Not a fan of this comments. If someone changes downvotes mapper he will 99% forget to change this comment also

Reply:
Fair point, but I currently don't have a better solution. Really don't want to iterate over the whole mapper at every vote/downvote.